### ISSUE TITLE
3D vs 5D array layout

### ISSUE THREAD

**Issue by tischi** [tischi/i2k-2020-s3-zarr-workshop#10]
I guess the arrays in bdv.n5 are 3D, right? I would assume so because in BDV the channel typically is encoded in the setup  and the timepoint in the timepoint group, leaving only x,y,z as dimensions for the data arrays. 
@constantinpape Is this correct?

In ome.zarr the arrays are 5D, right? 

@joshmoore, do you know whether this mapping from (3D array, `setup`, `timepoint`) (BDV) to 5D array (ome.zarr) is already taken care of somewhere in one of the classes? 

Speaking of this, I wonder whether we could in fact map `numChannels` (ome.zarr) to `numSetups` (bdv)?
Maybe we even have to? Because how would we otherwise open multi-channel files with the current infrastructure?

**Comment by constantinpape** [comment]
> I guess the arrays in bdv.n5 are 3D, right? I would assume so because in BDV the channel typically is encoded in the setup and the timepoint in the timepoint group, leaving only x,y,z as dimensions for the data arrays.
> @constantinpape Is this correct?

Yes, that's correct.

> In ome.zarr the arrays are 5D, right?

Yes, that's also correct. Note that in the data we have here, `t` and `c` are singleton dimensions, so the shape is always `(1, 1, x, y, z)`. So (for now) you could just pass a 3d view to the data.

**Comment by joshmoore** [comment]
`(1, 1, z, y, x)` but yes. Unfortunately, I don't think I have any sample Java code for that migration. The python is in https://github.com/tischi/i2k-2020-s3-zarr-workshop/issues/1#issuecomment-726046148

**Comment by constantinpape** [comment]
> `(1, 1, z, y, x)` but yes. 

Yes, good point.

> Unfortunately, I don't think I have any sample Java code for that migration. The python is in [#1 (comment)](https://github.com/tischi/i2k-2020-s3-zarr-workshop/issues/1#issuecomment-726046148)

We have done the migration already. I think the first issue for @tischi  is how to pass this to bdv, which expects 3d.
I would hope that there is some imglib function to pass a view that squeezes away the two singleton dimensions, but you would obviously know better than me...

**Comment by tischi** [comment]
> I would hope that there is some imglib function to pass a view that squeezes away the two singleton dimensions, but you would obviously know better than me...

;-) I think i'll manage somehow, but before hacking away, I just wanted to double-check with you guys that I am not missing something.

**Comment by tischi** [comment]
Ah, and maybe this got overlooked, so I take the liberty to repeat myself:

Speaking of this, I wonder whether we could in fact **map numChannels (ome.zarr) to numSetups (bdv)**?
Maybe we even have to? Because how would we otherwise open multi-channel files with the current infrastructure?

What do you think?

**Comment by constantinpape** [comment]
> Speaking of this, I wonder whether we could in fact **map numChannels (ome.zarr) to numSetups (bdv)**?

Maybe. I think we right now have at least two potential options on how multiple setups could be stored in ome.zarr:
- as multiple groups which contain separate images
- as a single group which contains an image with multiple channels
- (and this could also be mixed)

So I think before we decide how to deal with this, we need to understand these two different options more.

> Maybe we even have to? Because how would we otherwise open multi-channel files with the current infrastructure?

Yes, with the current bdv/MoBIE infrastructure that is the only way I can see too.

> What do you think?

Luckily we only have the case of a single channel and a single image per zarr root right now, so we don't need to
deal with multiple setup ids yet. So I think for the prototype it's fine to hard-code the number of setups to 1.

Longer term, I think we need to understand the different options I mentioned above.

**Comment by tischi** [comment]
@constantinpape @joshmoore 

I am making progress 🚀 

![image](https://user-images.githubusercontent.com/2157566/100074682-56434a80-2e3f-11eb-8db6-5562fb8bbf82.png)

When I zoom out and it is fetching the lower resolutions there is still something wrong, but I am on it.
![image](https://user-images.githubusercontent.com/2157566/100074615-4297e400-2e3f-11eb-8a98-4de5d638b9da.png)

**Comment by tischi** [comment]
@constantinpape 
Any chance the bug with the lower resolutions in on "your side"?

**Comment by constantinpape** [comment]
I wrote [a script to check the individual scale layers in napari](https://github.com/constantinpape/i2k-2020-s3-zarr-workshop/blob/main/check_data.py) and the result looks good:

You can see all the different scales (s0 in white, s1 in red, etc) roughly aligned when resized to the same size. 
So I don't think there is an issue with the data.

![Screenshot from 2020-11-24 11-36-22](https://user-images.githubusercontent.com/4263537/100083401-b808b200-2e49-11eb-92a7-fc696d13ddcc.png)

**Comment by tischi** [comment]
ok! and just to be sure they are all the same data type? unit8?

**Comment by joshmoore** [comment]
In the original, they definitely should be. Which dataset is this?  (i.e. s3.embl.de URL)

**Comment by constantinpape** [comment]
Yes, all the scale arrays are uint8. 

@joshmoore it's the following path on the embl s3:  `i2k-2020/prospr-myosin.ome.zarr`.

@tischi For debugging, could you write a script that just opens the individual scale arrays in bdv and check if these are loaded correctly?

**Comment by tischi** [comment]
> For debugging, could you write a script that just opens the individual scale arrays in bdv and check if these are loaded correctly?

My plan was to open the exact same file from n5.s3 and see whether I can spot any differences...

**Comment by tischi** [comment]
>  could you write a script that just opens the individual scale arrays in bdv and check if these are loaded correctly?

I am also not sure whether we have the code in Java yet to do this. I would need an equivalent of your python code:

```
with zarr.open(path) as f:
    im = f[scale][:]
```

like this

```
reader = new ZarrReader( path );
ImagePlus imp = reader.open( scale );
// or RandomAccessibleInterval imp = reader.open( scale );
```

@joshmoore does such ^ code exist already for Java? Would be cool for the workshop!!

**Comment by joshmoore** [comment]
Not that I know of, no, since I I've not yet tried to combine n5-imglib2 or n5-ij with n5-zarr. It's using a different underlying library, but in case it's useful https://github.com/dgault/ZarrReader/tree/prototype has now been newly minted.

**Comment by tischi** [comment]
@joshmoore 
I don't think it matters if the underlying library is different.
1. Is there an example of how to use the ZarrReader prototype?
1. Can I already add this as maven dependencies, or do I have to built it myself?

**Comment by tischi** [comment]
@constantinpape 
Not sure it matters, but I found some differences. 
The number of elements in the data blocks that are read sometimes is slightly different, e.g.

```
bdv.n5: setup0/timepoint0/s1 [1, 0, 0] 1998848
ome.zarr: s1 [1, 0, 0] 2097152
```
Could it be that we use different compression algorithms? 
Do we compress at all for ome.zarr?

**Comment by tischi** [comment]
@constantinpape 
Could it be that for the bdv.n5 the image boundary chunks are saved "cropped", only including within bounds pixels, whereas for ome.zarr they are currently saved non-cropped, including out-of-bounds pixels?

EDIT: I think that could explain why you can open it in Napari, which, I guess, looks at the data as 3D arrays, whereas in BDV the chunks are handed over as 1D arrays and then read out assuming a potentially wrong block size.

**Comment by joshmoore** [comment]
> Could it be that for the bdv.n5 the image boundary chunks are saved "cropped"

That is definitely one of the differences between n5 and zarr.

> * Is there an example of how to use the ZarrReader prototype?

Not yet, I think.

> * Can I already add this as maven dependencies, or do I have to built it myself?

The build is currently being worked on.

**Comment by constantinpape** [comment]
> Could it be that we use different compression algorithms?

Yes, n5 uses gzip, zarr uses blosc + lz4. This is because the default compression algorithms are different.

> Could it be that for the bdv.n5 the image boundary chunks are saved "cropped", only including within bounds pixels, whereas for ome.zarr they are currently saved non-cropped, including out-of-bounds pixels?

Yes, that is correct. This is a difference between the n5 and zarr specs.

> EDIT: I think that could explain why you can open it in Napari, which, I guess, looks at the data as 3D arrays, whereas in BDV the chunks are handed over as 1D arrays and then read out assuming a potentially wrong block size.

No, when reading via `zarr-python` the overhanging chunks are automatically cropped.
I guess this is not happening in `n5-zarr`.

**Comment by tischi** [comment]
> That is definitely one of the differences between n5 and zarr.

Ok, that's probably the problem. Talked with Constantin and I am trying to make use of SaalFeld's [N5CellLoader]( https://github.com/saalfeldlab/n5-imglib2/blob/master/src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5CellLoader.java#L129).
It's pretty tough as there is a solid mix of lambdas, functional programming, generics and type casting madness all combined, clearly beyond my comfort zone, but I am working on it and am still optimistic 👍

**Comment by tischi** [comment]
🥳 🎆 
And I even think I simplified the code a bit although I have not clue what I doing.

![image](https://user-images.githubusercontent.com/2157566/100127285-3df32000-2e7f-11eb-837f-00cbc07b9d67.png)

**Comment by tischi** [comment]
@constantinpape 
Could you please also upload the other two data sets?

**Comment by constantinpape** [comment]
> @constantinpape
> Could you please also upload the other two data sets?

Will do later and let you know once it's there.

**Comment by constantinpape** [comment]
@tischi I uploaded all the data to s3 and made a pr with the updated xmls #12.
